patches/api/0387-Add-damage-source-wrapper.patch

```diff
@@ -0,0 +1,1070 @@
++        return entitySource("thorns", damager).setMagic(true).setThorns(true).build();
++    }
++
++    @NotNull
++    static DamageOrigin ofExplosion() {
++        return ofExplosion(null);
++    }
++
++    @NotNull
++    static DamageOrigin ofExplosion(@Nullable LivingEntity source) {
++        if(source != null) {
++            return entitySource("explosion.player", source).setExplosion(true).scalesWithDifficulty(true).build();
++        }
++        return of("explosion").setExplosion(true).scalesWithDifficulty(true).build();
++    }
++
++    @NotNull
++    static DamageOrigin ofSonicBoom(@NotNull Entity damager) {
++        return entitySource("sonic_boom", damager).setMagic(true).bypassOf(DamageBypass.ARMOR, DamageBypass.ENCHANTMENT).build();
++    }
++
++    boolean isSweep();
++
++    boolean isCritical();
++
++    boolean isFromProjectile();
++
++    boolean isFromExplosion();
++
++    boolean isFromFire();
++
++    boolean isFromMagic();
++
++    boolean isFromFalling();
++
++    boolean willNotAggro();
++
++    boolean willScalesWithDifficulty();
++
++    default boolean canBypass(@NotNull DamageBypass... damageType) {
++        for(DamageBypass type : damageType) {
++            if(!canBypass(type)) return false;
++        }
++        return true;
++    }
++
++    boolean canBypass(@NotNull DamageBypass damageType);
++
++    boolean applyArmorExhaustion();
++
++    boolean applyMagicExhaustion();
++
++    boolean canDamageHelmetArmor();
++
++    float getFoodExhaustion();
++
++    @Nullable
++    Component getDeathMessageFor(@NotNull LivingEntity entity);
++
++    @NotNull
++    String getName();
++
++    boolean isSimilar(@NotNull DamageOrigin origin);
++}
+diff --git a/src/main/java/io/papermc/paper/entity/damage/EntityDamageOrigin.java b/src/main/java/io/papermc/paper/entity/damage/EntityDamageOrigin.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..59c0f896e2cdce43c2bc325702625882d9a538a4
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/entity/damage/EntityDamageOrigin.java
+@@ -0,0 +1,17 @@
++package io.papermc.paper.entity.damage;
++
++import org.bukkit.entity.Entity;
++import org.jetbrains.annotations.NotNull;
++import org.jetbrains.annotations.Nullable;
++
++public interface EntityDamageOrigin extends DamageOrigin {
++
++    boolean isThorns();
++
++    @Nullable
++    Entity getDirectSource();
++
++    @Nullable
++    Entity getSource();
++
++}
+diff --git a/src/main/java/io/papermc/paper/entity/damage/IndirectEntityDamageOrigin.java b/src/main/java/io/papermc/paper/entity/damage/IndirectEntityDamageOrigin.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..72e286e9c7c83ab767a3c0f704d4a681c20eede2
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/entity/damage/IndirectEntityDamageOrigin.java
+@@ -0,0 +1,4 @@
++package io.papermc.paper.entity.damage;
++
++public interface IndirectEntityDamageOrigin extends EntityDamageOrigin {
++}
```

Comment:
Is getDirectSource ever not null when it's just an EntityDamageOrigin?
I really think it should just be in IndirectEntityDamageOrigin here, and if vanilla ever decides that it used elsewhere we can simply move the method down.

Reply:
It could be worth it, vanilla just redirect to the classic source for direct entity damage source